### ISSUE TITLE
inconsistent gas-usage in `create-rollapp` command

### ISSUE THREAD

**Issue by anhductn2001** [dymensionxyz/dymension#1304]
Exp log:
- 1st send tx:
```
dymd q tx 780D7D6A66838E5B44CC3C330375295B1B5EC8F77C7AA0E5702EBD06D29A6E78
code: 11
codespace: sdk
data: ""
events:
- attributes:
  - index: true
    key: fee
    value: ""
  type: tx
- attributes:
  - index: true
    key: acc_seq
    value: dym1rh46rgnlvlg6sp7vajql7ugw58wglz9w7cacg7/0
  type: tx
- attributes:
  - index: true
    key: [redacted:possible-secret]
    value: hBZL1nDpZpFRPryTRFd9kX/mLONIjX5PhJv1XHNlj/IOUvS3xNfvbmFdzhYMAY4gUKhHPjFbQCjlYj1h5fO+UAE=
  type: tx
gas_used: "200595"
gas_wanted: "200000"
height: "15"
info: ""
logs: []
raw_log: 'out of gas in location: ReadFlat; gasWanted: 200000, gasUsed: 200595: out
  of gas'
timestamp: "2024-10-10T15:27:54Z"
tx:
  '@type': /cosmos.tx.v1beta1.Tx
  auth_info:
    fee:
      amount: []
      gas_limit: "200000"
      granter: ""
      payer: ""
    signer_infos:
    - mode_info:
        single:
          mode: SIGN_MODE_DIRECT
      public_key:
        '@type': /ethermint.crypto.v1.ethsecp256k1.PubKey
        key: [redacted:possible-secret]
      sequence: "0"
    tip: null
  body:
    extension_options: []
    memo: ""
    messages:
    - '@type': /dymensionxyz.dymension.rollapp.MsgCreateRollapp
      alias: icxuv
      creator: dym1rh46rgnlvlg6sp7vajql7ugw58wglz9w7cacg7
      genesis_info:
        bech32_prefix: ethm
        genesis_accounts:
          accounts:
          - address: dym1c0jgquy7aankvn8ruetdmc44cdgc3x55f2wnzm
            amount: "100000000000000000000"
        genesis_checksum: aaa
        initial_supply: "100000000010100000000000000000000"
        native_denom:
          base: urax
          display: rax
          exponent: 18
        sealed: false
      initial_sequencer: dym1rh46rgnlvlg6sp7vajql7ugw58wglz9w7cacg7
      metadata:
        description: This is a description of the Rollapp.
        display_name: ""
        explorer_url: ""
        fee_denom: null
        genesis_url: ""
        logo_url: ""
        tagline: ""
        telegram: https://t.me/example
        website: https://dymension.xyz/
        x: https://x.com/dymension
      rollapp_id: rollappevm_1234-1
      vm_type: EVM
    non_critical_extension_options: []
    timeout_height: "0"
  signatures:
  - hBZL1nDpZpFRPryTRFd9kX/mLONIjX5PhJv1XHNlj/IOUvS3xNfvbmFdzhYMAY4gUKhHPjFbQCjlYj1h5fO+UAE=
txhash: 780D7D6A66838E5B44CC3C330375295B1B5EC8F77C7AA0E5702EBD06D29A6E78
```

- 2nd send tx:
```
dymd q tx E2A1634E6B4821F9E3E3DFC363649D3AAFEEE905F8C65983A3C23C80668D1830
code: 0
codespace: ""
data: 123A0A382F64796D656E73696F6E78797A2E64796D656E73696F6E2E726F6C6C6170702E4D7367437265617465526F6C6C617070526573706F6E7365
events:
- attributes:
  - index: true
    key: fee
    value: ""
  type: tx
- attributes:
  - index: true
    key: acc_seq
    value: dym1rh46rgnlvlg6sp7vajql7ugw58wglz9w7cacg7/1
  type: tx
- attributes:
  - index: true
    key: [redacted:possible-secret]
    value: MKQOekHCcPdteLCmPE/cZwcP720k5Sgvo43xLNzeT3MeAsgc1RCt4Kkmd+YVOrNNg1Hf5lJpD9DDLrOrysk5ywE=
  type: tx
- attributes:
  - index: true
    key: action
    value: /dymensionxyz.dymension.rollapp.MsgCreateRollapp
  - index: true
    key: sender
    value: dym1rh46rgnlvlg6sp7vajql7ugw58wglz9w7cacg7
  - index: true
    key: module
    value: dymension
  type: message
- attributes:
  - index: true
    key: spender
    value: dym1rh46rgnlvlg6sp7vajql7ugw58wglz9w7cacg7
  - index: true
    key: amount
    value: 250000000000000000000adym
  type: coin_spent
- attributes:
  - index: true
    key: [redacted:possible-secret]
    value: dym1kmqwc8wf88vukz9fpgkz5yj4n83jlexlvatzs4
  - index: true
    key: amount
    value: 250000000000000000000adym
  type: coin_received
- attributes:
  - index: true
    key: [redacted:possible-secret]
    value: dym1kmqwc8wf88vukz9fpgkz5yj4n83jlexlvatzs4
  - index: true
    key: sender
    value: dym1rh46rgnlvlg6sp7vajql7ugw58wglz9w7cacg7
  - index: true
    key: amount
    value: 250000000000000000000adym
  type: transfer
- attributes:
  - index: true
    key: sender
    value: dym1rh46rgnlvlg6sp7vajql7ugw58wglz9w7cacg7
  type: message
- attributes:
  - index: true
    key: spender
    value: dym1kmqwc8wf88vukz9fpgkz5yj4n83jlexlvatzs4
  - index: true
    key: amount
    value: 250000000000000000000adym
  type: coin_spent
- attributes:
  - index: true
    key: burner
    value: dym1kmqwc8wf88vukz9fpgkz5yj4n83jlexlvatzs4
  - index: true
    key: amount
    value: 250000000000000000000adym
  type: burn
- attributes:
  - index: true
    key: [redacted:possible-secret]
    value: Alias
  - index: true
    key: name
    value: icxuv
  - index: true
    key: price
    value: 250000000000000000000adym
  - index: true
    key: buyer
    value: rollappevm_1234-1
  type: dymns_sell
- attributes:
  - index: true
    key: alias
    value: icxuv
  - index: true
    key: creator
    value: dym1rh46rgnlvlg6sp7vajql7ugw58wglz9w7cacg7
  - index: true
    key: [redacted:possible-secret]
    value: '{"genesis_checksum":"aaa","bech32_prefix":"ethm","native_denom":{"display":"rax","base":"urax","exponent":18},"initial_supply":"100000000010100000000000000000000","sealed":false,"genesis_accounts":{"accounts":[{"amount":"100000000000000000000","address":"dym1c0jgquy7aankvn8ruetdmc44cdgc3x55f2wnzm"}]}}'
  - index: true
    key: [redacted:possible-secret]
    value: dym1rh46rgnlvlg6sp7vajql7ugw58wglz9w7cacg7
  - index: true
    key: [redacted:possible-secret]
    value: '{"website":"https://dymension.xyz/","description":"This is a description
      of the Rollapp.","logo_url":"","telegram":"https://t.me/example","x":"https://x.com/dymension","genesis_url":"","display_name":"","tagline":"","explorer_url":"","fee_denom":null}'
  - index: true
    key: [redacted:possible-secret]
    value: rollappevm_1234-1
  - index: true
    key: vm_type
    value: EVM
  type: dymensionxyz.dymension.rollapp.MsgCreateRollapp
gas_used: "197088"
gas_wanted: "200000"
height: "21"
info: ""
logs:
- events:
  - attributes:
    - key: action
      value: /dymensionxyz.dymension.rollapp.MsgCreateRollapp
    - key: sender
      value: dym1rh46rgnlvlg6sp7vajql7ugw58wglz9w7cacg7
    - key: module
      value: dymension
    type: message
  - attributes:
    - key: spender
      value: dym1rh46rgnlvlg6sp7vajql7ugw58wglz9w7cacg7
    - key: amount
      value: 250000000000000000000adym
    type: coin_spent
  - attributes:
    - key: [redacted:possible-secret]
      value: dym1kmqwc8wf88vukz9fpgkz5yj4n83jlexlvatzs4
    - key: amount
      value: 250000000000000000000adym
    type: coin_received
  - attributes:
    - key: [redacted:possible-secret]
      value: dym1kmqwc8wf88vukz9fpgkz5yj4n83jlexlvatzs4
    - key: sender
      value: dym1rh46rgnlvlg6sp7vajql7ugw58wglz9w7cacg7
    - key: amount
      value: 250000000000000000000adym
    type: transfer
  - attributes:
    - key: sender
      value: dym1rh46rgnlvlg6sp7vajql7ugw58wglz9w7cacg7
    type: message
  - attributes:
    - key: spender
      value: dym1kmqwc8wf88vukz9fpgkz5yj4n83jlexlvatzs4
    - key: amount
      value: 250000000000000000000adym
    type: coin_spent
  - attributes:
    - key: burner
      value: dym1kmqwc8wf88vukz9fpgkz5yj4n83jlexlvatzs4
    - key: amount
      value: 250000000000000000000adym
    type: burn
  - attributes:
    - key: [redacted:possible-secret]
      value: Alias
    - key: name
      value: icxuv
    - key: price
      value: 250000000000000000000adym
    - key: buyer
      value: rollappevm_1234-1
    type: dymns_sell
  - attributes:
    - key: alias
      value: icxuv
    - key: creator
      value: dym1rh46rgnlvlg6sp7vajql7ugw58wglz9w7cacg7
    - key: [redacted:possible-secret]
      value: '{"genesis_checksum":"aaa","bech32_prefix":"ethm","native_denom":{"display":"rax","base":"urax","exponent":18},"initial_supply":"100000000010100000000000000000000","sealed":false,"genesis_accounts":{"accounts":[{"amount":"100000000000000000000","address":"dym1c0jgquy7aankvn8ruetdmc44cdgc3x55f2wnzm"}]}}'
    - key: [redacted:possible-secret]
      value: dym1rh46rgnlvlg6sp7vajql7ugw58wglz9w7cacg7
    - key: [redacted:possible-secret]
      value: '{"website":"https://dymension.xyz/","description":"This is a description
        of the Rollapp.","logo_url":"","telegram":"https://t.me/example","x":"https://x.com/dymension","genesis_url":"","display_name":"","tagline":"","explorer_url":"","fee_denom":null}'
    - key: [redacted:possible-secret]
      value: rollappevm_1234-1
    - key: vm_type
      value: EVM
    type: dymensionxyz.dymension.rollapp.MsgCreateRollapp
  log: ""
  msg_index: 0
raw_log: '[{"msg_index":0,"events":[{"type":"message","attributes":[{"key":"action","value":"/dymensionxyz.dymension.rollapp.MsgCreateRollapp"},{"key":"sender","value":"dym1rh46rgnlvlg6sp7vajql7ugw58wglz9w7cacg7"},{"key":"module","value":"dymension"}]},{"type":"coin_spent","attributes":[{"key":"spender","value":"dym1rh46rgnlvlg6sp7vajql7ugw58wglz9w7cacg7"},{"key":"amount","value":"250000000000000000000adym"}]},{"type":"coin_received","attributes":[{"key":"[redacted:possible-secret]","value":"dym1kmqwc8wf88vukz9fpgkz5yj4n83jlexlvatzs4"},{"key":"amount","value":"250000000000000000000adym"}]},{"type":"transfer","attributes":[{"key":"[redacted:possible-secret]","value":"dym1kmqwc8wf88vukz9fpgkz5yj4n83jlexlvatzs4"},{"key":"sender","value":"dym1rh46rgnlvlg6sp7vajql7ugw58wglz9w7cacg7"},{"key":"amount","value":"250000000000000000000adym"}]},{"type":"message","attributes":[{"key":"sender","value":"dym1rh46rgnlvlg6sp7vajql7ugw58wglz9w7cacg7"}]},{"type":"coin_spent","attributes":[{"key":"spender","value":"dym1kmqwc8wf88vukz9fpgkz5yj4n83jlexlvatzs4"},{"key":"amount","value":"250000000000000000000adym"}]},{"type":"burn","attributes":[{"key":"burner","value":"dym1kmqwc8wf88vukz9fpgkz5yj4n83jlexlvatzs4"},{"key":"amount","value":"250000000000000000000adym"}]},{"type":"dymns_sell","attributes":[{"key":"[redacted:possible-secret]","value":"Alias"},{"key":"name","value":"icxuv"},{"key":"price","value":"250000000000000000000adym"},{"key":"buyer","value":"rollappevm_1234-1"}]},{"type":"dymensionxyz.dymension.rollapp.MsgCreateRollapp","attributes":[{"key":"alias","value":"icxuv"},{"key":"creator","value":"dym1rh46rgnlvlg6sp7vajql7ugw58wglz9w7cacg7"},{"key":"[redacted:possible-secret]","value":"{\"genesis_checksum\":\"aaa\",\"bech32_prefix\":\"ethm\",\"native_denom\":{\"display\":\"rax\",\"base\":\"urax\",\"exponent\":18},\"initial_supply\":\"100000000010100000000000000000000\",\"sealed\":false,\"genesis_accounts\":{\"accounts\":[{\"amount\":\"100000000000000000000\",\"address\":\"dym1c0jgquy7aankvn8ruetdmc44cdgc3x55f2wnzm\"}]}}"},{"key":"[redacted:possible-secret]","value":"dym1rh46rgnlvlg6sp7vajql7ugw58wglz9w7cacg7"},{"key":"[redacted:possible-secret]","value":"{\"website\":\"https://dymension.xyz/\",\"description\":\"This
  is a description of the Rollapp.\",\"logo_url\":\"\",\"telegram\":\"https://t.me/example\",\"x\":\"https://x.com/dymension\",\"genesis_url\":\"\",\"display_name\":\"\",\"tagline\":\"\",\"explorer_url\":\"\",\"fee_denom\":null}"},{"key":"[redacted:possible-secret]","value":"rollappevm_1234-1"},{"key":"vm_type","value":"EVM"}]}]}]'
timestamp: "2024-10-10T15:28:06Z"
tx:
  '@type': /cosmos.tx.v1beta1.Tx
  auth_info:
    fee:
      amount: []
      gas_limit: "200000"
      granter: ""
      payer: ""
    signer_infos:
    - mode_info:
        single:
          mode: SIGN_MODE_DIRECT
      public_key:
        '@type': /ethermint.crypto.v1.ethsecp256k1.PubKey
        key: [redacted:possible-secret]
      sequence: "1"
    tip: null
  body:
    extension_options: []
    memo: ""
    messages:
    - '@type': /dymensionxyz.dymension.rollapp.MsgCreateRollapp
      alias: icxuv
      creator: dym1rh46rgnlvlg6sp7vajql7ugw58wglz9w7cacg7
      genesis_info:
        bech32_prefix: ethm
        genesis_accounts:
          accounts:
          - address: dym1c0jgquy7aankvn8ruetdmc44cdgc3x55f2wnzm
            amount: "100000000000000000000"
        genesis_checksum: aaa
        initial_supply: "100000000010100000000000000000000"
        native_denom:
          base: urax
          display: rax
          exponent: 18
        sealed: false
      initial_sequencer: dym1rh46rgnlvlg6sp7vajql7ugw58wglz9w7cacg7
      metadata:
        description: This is a description of the Rollapp.
        display_name: ""
        explorer_url: ""
        fee_denom: null
        genesis_url: ""
        logo_url: ""
        tagline: ""
        telegram: https://t.me/example
        website: https://dymension.xyz/
        x: https://x.com/dymension
      rollapp_id: rollappevm_1234-1
      vm_type: EVM
    non_critical_extension_options: []
    timeout_height: "0"
  signatures:
  - MKQOekHCcPdteLCmPE/cZwcP720k5Sgvo43xLNzeT3MeAsgc1RCt4Kkmd+YVOrNNg1Hf5lJpD9DDLrOrysk5ywE=
txhash: E2A1634E6B4821F9E3E3DFC363649D3AAFEEE905F8C65983A3C23C80668D1830
```

## AI triage
Re-triage complete. Here is what I found and the verdict.

**TLDR:** The issue (Oct 2024) reports `create-rollapp` failing out-of-gas at the CLI's default 200k gas limit, plus repeated `--dry-run` simulations of the identical tx toggling between two gas estimates. The human update (mtsitrin, confirmed by anhductn2001, Dec 2024) established that the default gas limit caused the failure and `--gas auto` works around it — the remaining work is root-causing the non-deterministic estimation. I verified read-only that nothing has landed since: the issue is open with no labels and no linked PRs, the sibling issue #867 ("Gas estimation is inconsistent") is also still open, and the git log since Oct 2024 shows no simulation-determinism fix. Meanwhile the codebase has moved substantially (cosmos-sdk is now v0.50.14, and the create-rollapp hook path now spans dymns alias registration with txfees conversion, streamer/incentives gauge creation, and sponsorship endorsement), so the first step of any fix is reproducing the variance at HEAD on a local devnet — the original test environment no longer exists. That reproduction, root-cause, and fix is fully local work needing no credentials or human hands, so it dispatches; it touches consensus-adjacent gas metering on the settlement layer, so it routes critical.

---

## Research summary

Original report (anhductn2001, Oct 2024, migration testnet): a `MsgCreateRollapp` sent with the cosmos-sdk CLI's default gas limit of 200000 failed with `out of gas in location: ReadFlat; gasWanted: 200000, gasUsed: 200595`. An immediately re-sent identical tx (next sequence) succeeded using 197088 gas. Note the failing tx was the account's first ever (sequence 0), and first-use txs consume extra gas persisting the account pubkey — consistent with the first attempt landing just over the default limit while the retry fit under it.

Per mtsitrin's update on the thread (Dec 2024): the failure happened because the tx used the default gasLimit 200000; using `--gas auto` solves it; but four consecutive `--dry-run` simulations of the same tx returned estimates alternating between two values, 194803 and 197595 (pattern a, b, b, a — suggesting state- or height-dependent variance rather than pure randomization). Per anhductn2001's reply: `--gas auto` confirmed working; the difference between identical simulations still needs investigation. mtsitrin also linked #867 as possibly related.

Already done (per the thread): diagnosis of the failure cause (default gas limit) and confirmation of the `--gas auto` workaround — both mtsitrin and anhductn2001. Remaining and unowned: root-causing the non-deterministic gas estimation; no one is assigned, no PR exists, and #867 remains open with no fix either — the worker dispatched on this brief takes that remaining step.

Confirmed from the clone at HEAD (commit `945d7361`, Aug 2026): no gas-determinism fix has landed. The `CreateRollapp` handler now runs `CheckIfRollappExists`, min-bond validation, `SetRollapp`, then the `RollappCreated` multi-hook — dymns alias registration (which prices the alias, converts the cost via the txfees keeper, and burns it) and the streamer hook (creates a rollapp gauge and saves a sponsorship endorsement). The repo is on cosmos-sdk v0.50.14 with dymension forks of ethermint and osmosis v15 — a major platform upgrade since the report, so the 2024 gas numbers and possibly the variance itself are stale; the behavior must be re-established at HEAD. The original RPC endpoint was a since-decommissioned migration testnet; reproduction is local-devnet work.

## Root cause hypothesis

Two distinct problems are folded into this issue:

1. **The out-of-gas failure** — not an in-repo bug: the cosmos-sdk CLI default `--gas 200000` sat marginally below actual `create-rollapp` consumption at the time, and the first-tx-from-account pubkey persistence pushed the first attempt over the line. `--gas auto` is the correct usage; confirmed by both reporters.
2. **The two-valued simulation toggle** — unpinned. The a, b, b, a pattern across seconds points at per-block state changes altering the number or size of store reads during simulation (candidate paths: params/pool/gauge state read during the dymns hook's txfees cost conversion, or simulate-mode ante differences such as pubkey simulation), with Go-map iteration order changing store-access counts as a secondary candidate. If any such variance existed in the deliver (non-simulate) path it would be a consensus-split risk, which is why this must be pinned rather than shrugged off. It is also possible the variance no longer reproduces on cosmos-sdk v0.50.14 — the fix then is the evidence plus a regression test, not a code change.

## Areas involved

- `x/rollapp` msg server (`CreateRollapp`) and its `MultiRollappHooks` chain: `x/dymns` alias-registration hook (including the txfees `CalcBaseInCoin` conversion), `x/streamer`/`x/incentives` rollapp gauge creation, `x/sponsorship` endorsement.
- The simulate/ante path (cosmos-sdk v0.50.14, dymension ethermint fork) as exercised by tx simulation.
- Test harness: app-level or keeper-level Go test driving repeated gas-metered simulation of an identical `MsgCreateRollapp` against identical committed state.

## Acceptance criteria

- A regression test checked into this repo that executes gas-metered simulation of an identical `MsgCreateRollapp` at least 20 times against identical committed state and asserts every run yields the same gas figure.
- If variance reproduces at HEAD: the mechanism is identified and fixed in this repo, and the regression test proves determinism after the fix. Pinned invariant: gas consumption must be a pure function of tx bytes and committed state in both simulate and deliver modes. A fix that alters deliver-path gas for existing tx types is consensus-breaking and ships through the repo's upgrade-handler conventions, stated explicitly in the PR.
- If the mechanism lives in a dymensionxyz-maintained fork dependency (ethermint, osmosis) rather than this repo: the PR still lands the regression test, documents the exact upstream mechanism with reproducing evidence, and names the fork change needed — that documented handoff is the deliverable for the upstream half.
- If no variance reproduces at HEAD: the PR lands the determinism test with the recorded N-run evidence, and states that the historical failure was the CLI default gas limit with `--gas auto` as correct usage — that evidence resolves the issue.
- The test runs green in CI alongside the existing suite; as critical work, the author runs every CI-gated suite locally before the PR exists.

## Open questions

None — reproduction steps are in the issue, and the expected behavior (identical simulations yield identical gas) is unambiguous.

---

**Comment by mtsitrin** [comment]
@anhductn2001 are u using `--gas auto` or fixed gas?

**Comment by mtsitrin** [comment]
Edit:
* issue happend as tx uses gasLimit 200000 by default
* using `gas auto` solves this

But it seems like gas consumption is not consistent:
```
dymd --node https://rpc-dym-migration-test-2.mzonder.com:443 tx rollapp create-rollapp ndfotoken_12344-1 notokfdsadas1 EVM --keyring-backend test --keyring-dir ~/.rollapp_evm/sequencer_keys/ --from dym1tnns80tngq22g3zzzpufl4225hwxle3p5958k6 --fees 1dym --dry-run
gas estimate: 194803
dymd --node https://rpc-dym-migration-test-2.mzonder.com:443 tx rollapp create-rollapp ndfotoken_12344-1 notokfdsadas1 EVM --keyring-backend test --keyring-dir ~/.rollapp_evm/sequencer_keys/ --from dym1tnns80tngq22g3zzzpufl4225hwxle3p5958k6 --fees 1dym --dry-run
gas estimate: 197595
dymd --node https://rpc-dym-migration-test-2.mzonder.com:443 tx rollapp create-rollapp ndfotoken_12344-1 notokfdsadas1 EVM --keyring-backend test --keyring-dir ~/.rollapp_evm/sequencer_keys/ --from dym1tnns80tngq22g3zzzpufl4225hwxle3p5958k6 --fees 1dym --dry-run
gas estimate: 197595
dymd --node https://rpc-dym-migration-test-2.mzonder.com:443 tx rollapp create-rollapp ndfotoken_12344-1 notokfdsadas1 EVM --keyring-backend test --keyring-dir ~/.rollapp_evm/sequencer_keys/ --from dym1tnns80tngq22g3zzzpufl4225hwxle3p5958k6 --fees 1dym --dry-run
gas estimate: 194803
```

we can see for the same tx, the gas estimation is toggling between 2 different values

**Comment by anhductn2001** [comment]
yea, u right @mtsitrin, after adding `--gas auto` it worked fine, but we should also investigate why there is a difference between the 2  txs

**Comment by omriai1** [comment]
Update received — re-groomed and dispatching to the workforce.

**Comment by omriai1** [comment]
_Re-groom bookkeeping — updated in place, safe to ignore._

**Comment by omriai1** [comment]
Workforce worker dispatched (coder: opus, reviewer: sol).

**Comment by omriai1** [comment]
PR opened: #2211
